src/geometry_aux.cpp

```diff
@@ -40,6 +42,57 @@ void update_universe_cell_count(int32_t a, int32_t b)
   }
 }
 
+extern "C" void openmc_get_optical_thickness(Position start_voxel_min,
+  Position start_voxel_max, Position end_voxel_min, Position end_voxel_max,
+  int num_rays, double* output)
+{
+  double total_optical_thickness = 0.0;
+  SpatialBox start_box(start_voxel_min, start_voxel_max);
+  SpatialBox end_box(end_voxel_min, end_voxel_max);
+
+  int64_t id = 1;
+  uint64_t seed = init_seed(id, STREAM_SOURCE);
+
+  initialize_data();
+
+  for (int i = 0; i < num_rays; ++i) {
+    Position start_sampled_position = start_box.sample(&seed);
+    Position end_sampled_position = end_box.sample(&seed);
+
+    Position direction = (end_sampled_position - start_sampled_position) /
+                         (end_sampled_position - start_sampled_position).norm();
+
+    Particle p;
+    SourceSite site;
+    site.E = 1.0;
+    site.particle = ParticleType::neutron;
+    site.r = start_sampled_position;
+    site.u = direction;
+    p.from_source(&site);
+
+    double optical_thickness = 0.0;
+
+    while (true) {
+      if (!exhaustive_find_cell(p))
+        break;
+
+      BoundaryInfo boundary = distance_to_boundary(p);
+      double distance = boundary.distance;
+
+      p.event_calculate_xs();
+      optical_thickness += distance * p.macro_xs().total;
+
+      if (boundary.surface_index == -1)
```

Comment:
We need an issue to focus on the need to improve/correct this stopping condition.